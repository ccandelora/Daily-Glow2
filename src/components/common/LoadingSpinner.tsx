import React, { useEffect } from 'react';
import { Animated, StyleSheet, View, ViewStyle } from 'react-native';
import theme from '../../constants/theme';

interface LoadingSpinnerProps {
  size?: number;
  color?: string;
  style?: ViewStyle;
}

export const LoadingSpinner: React.FC<LoadingSpinnerProps> = ({
  size = 40,
  color = theme.COLORS.primary.green,
  style,
}) => {
  const rotateAnim = React.useRef(new Animated.Value(0)).current;

  useEffect(() => {
    const animation = Animated.loop(
      Animated.timing(rotateAnim, {
        toValue: 1,
        duration: 1000,
        useNativeDriver: true,
      })
    );
    animation.start();

    return () => animation.stop();
  }, [rotateAnim]);

  const spin = rotateAnim.interpolate({
    inputRange: [0, 1],
    outputRange: ['0deg', '360deg'],
  });

  return (
    <View style={[styles.container, style]}>
      <Animated.View
        style={[
          styles.spinner,
          {
            width: size,
            height: size,
            borderColor: color,
            transform: [{ rotate: spin }],
          },
        ]}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  spinner: {
    borderWidth: 3,
    borderRadius: theme.BORDER_RADIUS.circle,
    borderTopColor: 'transparent',
  },
}); 