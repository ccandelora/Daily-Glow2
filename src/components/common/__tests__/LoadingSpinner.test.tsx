import React from 'react';
import { render } from '@testing-library/react-native';
import { Animated } from 'react-native';
import { LoadingSpinner } from '../LoadingSpinner';

// Instead of mocking NativeAnimatedHelper, just mock the specific animation methods we need
jest.mock('react-native', () => {
  const rn = jest.requireActual('react-native');
  
  return {
    ...rn,
    Animated: {
      ...rn.Animated,
      timing: jest.fn(() => ({
        start: jest.fn(),
        stop: jest.fn(),
      })),
      loop: jest.fn(() => ({
        start: jest.fn(),
        stop: jest.fn(),
      })),
      Value: jest.fn(() => ({
        interpolate: jest.fn(() => 'interpolated-value'),
      })),
    },
  };
});

describe('LoadingSpinner Component', () => {
  test('renders correctly with default props', () => {
    const { toJSON } = render(<LoadingSpinner />);
    expect(toJSON()).toMatchSnapshot('default-spinner');
  });

  test('applies custom size', () => {
    const customSize = 60;
    const { toJSON } = render(<LoadingSpinner size={customSize} />);
    expect(toJSON()).toMatchSnapshot(`spinner-with-size-${customSize}`);
  });

  test('applies custom color', () => {
    const customColor = '#FF0000';
    const { toJSON } = render(<LoadingSpinner color={customColor} />);
    expect(toJSON()).toMatchSnapshot(`spinner-with-color-${customColor}`);
  });

  test('applies custom style', () => {
    const customStyle = { marginTop: 20, backgroundColor: '#F0F0F0' };
    const { toJSON } = render(<LoadingSpinner style={customStyle} />);
    expect(toJSON()).toMatchSnapshot('spinner-with-custom-style');
  });
}); 