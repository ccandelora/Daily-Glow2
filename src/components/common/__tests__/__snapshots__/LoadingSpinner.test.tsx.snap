// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingSpinner Component applies custom color: spinner-with-color-#FF0000 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    style={
      {
        "borderColor": "#FF0000",
        "borderRadius": 999,
        "borderTopColor": "transparent",
        "borderWidth": 3,
        "height": 40,
        "transform": [
          {
            "rotate": "0deg",
          },
        ],
        "width": 40,
      }
    }
  />
</View>
`;

exports[`LoadingSpinner Component applies custom size: spinner-with-size-60 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    style={
      {
        "borderColor": "#00C853",
        "borderRadius": 999,
        "borderTopColor": "transparent",
        "borderWidth": 3,
        "height": 60,
        "transform": [
          {
            "rotate": "0deg",
          },
        ],
        "width": 60,
      }
    }
  />
</View>
`;

exports[`LoadingSpinner Component applies custom style: spinner-with-custom-style 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "justifyContent": "center",
      },
      {
        "backgroundColor": "#F0F0F0",
        "marginTop": 20,
      },
    ]
  }
>
  <View
    collapsable={false}
    style={
      {
        "borderColor": "#00C853",
        "borderRadius": 999,
        "borderTopColor": "transparent",
        "borderWidth": 3,
        "height": 40,
        "transform": [
          {
            "rotate": "0deg",
          },
        ],
        "width": 40,
      }
    }
  />
</View>
`;

exports[`LoadingSpinner Component renders correctly with default props: default-spinner 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    style={
      {
        "borderColor": "#00C853",
        "borderRadius": 999,
        "borderTopColor": "transparent",
        "borderWidth": 3,
        "height": 40,
        "transform": [
          {
            "rotate": "0deg",
          },
        ],
        "width": 40,
      }
    }
  />
</View>
`;
